In [86]:
import os 
# This is to get the directory that the program  
# is currently running in. 
dir_path = os.path.dirname(os.path.realpath("C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\")) 
##print(dir_path)
filepdf = []
rootpdf = []
filedoc = []
rootdoc = []
filedocx = []
rootdocx = []
path = []
  
for root, dirs, files in os.walk(dir_path): 
    for file in files:  
  
        # change the extension from '.mp3' to  
        # the one of your choice. 
        if file.endswith('.pdf'):
            #print (root+'/'+str(file))
            a = root+'/'+str(file)
            filepdf.append(file)
            rootpdf.append(a)
            ##print(a)
        
        elif file.endswith('.doc'):
            #print (root+'/'+str(file))
            b = root+'/'+str(file)
            filedoc.append(file)
            rootdoc.append(b)
        
        else: 
            #print (root+'/'+str(file))
            c = root+'/'+str(file)
            filedocx.append(file)
            rootdocx.append(c)
       

In [87]:
import pandas as pd

In [88]:
rootpdf

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 2603 - Fall 2017 Syllabus (1).pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 2603 - Fall 2017 Syllabus.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MBA 3250Syllabus.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS 2627.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 msis 2629 fall2017.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2624-Syllabus-2017Fall-pdf (1).pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2624-Syllabus-2017Fall-pdf.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2641_Fall2017_Syllabus-Seidler (1).pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2641_Fall2017_Syllabu

In [89]:
rootdoc

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 41 syllabus Fall 2017A.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 41_3_Statistics Data Analysis II_Fall 2017_syllabus    9_18_2017.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 41_4_Statistics Data Analysis II_Fall 2017_syllabus    9_18_2017.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Spring 2018/Spring 2018 OMIS 3200 SCU Quantitative Methods_5 weeks_OMIS 3200_MW Spring 2018_Syllabus   5_1_2018.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Spring 2018/Spring 2018 OMIS 40-3  Syllabus_Statistics _ Data Analysis I_TTH 830_Syllabus_Spring 2018.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Spring 2018/Spring 2018 OMIS 40-4  Syllabus_Statistics _ Data Analysis I_TTH 1210_Syllabus_Spring 2018.doc',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\

In [90]:
rootdocx

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 40 Syllabus Fall 2017.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS 2602 syllabus_Fall 2017.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 2603 - Winter 2018 Syllabus.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 63945-MSIS2802-68943-IDIS3802-syllabus-Winter2018.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 63946-MSIS2802-68943-IDIS3802-syllabus-Winter2018.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 Escobar_OMIS 015_W18_R_Syllabus.docx',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 Escobar_OMIS 015_W18_T_Syllabus.docx',
 'C:\\User

**Write a function to check if the parameter exists in the path, if yes, return the path**

In [91]:
#def calcpath(param):
#    arr =[]
#    val = param
#    for root in rootpdf:
#        if val in root :
#            arr.append(root)
#    for root in rootdoc:
#        if val in root :
#            arr.append(root)
#    for root in rootdocx:
#        if val in root :
#            arr.append(root)
#    return arr

In [92]:
import PyPDF2 
import textract
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

In [138]:
from PyPDF2 import PdfFileWriter, PdfFileReader
from nltk.tokenize import sent_tokenize
def checkcrs(param, dep):
    arr = []
    val = str(param)
    dep = dep
    comb = dep+param
    for root in rootpdf:
        pdfFileObj = open(root, 'rb')
        #The pdfReader variable is a readable object that will be parsed.
        pdfReader = PyPDF2.PdfFileReader(pdfFileObj)
        #Discerning the number of pages will allow us to parse through all the pages.
        #num_pages = pdfReader.numPages
        num_pages = 1
        count = 0
        pdftext = ""
        #The while loop will read each page.
        while count < num_pages:
            pageObj = pdfReader.getPage(count)
            count +=1
            pdftext += pageObj.extractText()
        wordspdf = word_tokenize(pdftext)
        #wordspdf = pdftext.split()
        #print(wordspdf)
        if val in wordspdf:
            if dep in wordspdf:
                arr.append(root)
        elif comb in wordspdf:
            arr.append(root)
    return arr
            

In [118]:
dfcourse = pd.read_excel('Course Table.xls')

In [146]:
dfcourse.head()

Course ID Subject Catalog                          Descr Status   Eff Date  \
0      63018    ACTG     300           Financial Accounting      A 2008-01-24   
1      50122    ACTG     301    Actg for Business Decisions      A 1999-10-12   
2      63045    ACTG     302          Managerial Accounting      A 2009-10-08   
3      10049    ACTG     303  Corporate Financial Reporting      A 2009-10-08   
4      60566    ACTG     303  Corporate Financial Reporting      A 2007-12-10   

  Consent Grading  Min Units  Max Units Repeatable  Allowd Unt  Allow Comp  \
0       N     GRD          3          3          N           3           1   
1       N     GRD          3          3          N           3           1   
2       N     GRD          3          3          N           3           1   
3       N     GRD          3          3          N           3           1   
4       N     GRD          3          3          N           3           1   

  Allow Mult Campus Acad Org Career                     Long Title  \
0          N    NaN     ACTG   BUSN           Financial Accounting   
1          N    NaN     ACTG   BUSN    Actg for Business Decisions   
2          N    NaN     ACTG   BUSN          Managerial Accounting   
3          N    NaN     ACTG   BUSN  Corporate Financial Reporting   
4          N  OFCMP     ACTG   BUSN  Corporate Financial Reporting   

                                             Descr.1  
0  Introduces the roles, concepts, principles, le...  
1                                                NaN  
2  Focused study of issues relating to the measur...  
3                                                NaN  
4  An in-depth study of the concepts underlying e...

In [124]:
arrcourse = checkcrs('2601', 'MSIS')

In [125]:
arrcourse

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Summer 2018/MSIS2601-Outline.pdf']

In [126]:
arr3200 = checkcrs('3200','OMIS')

In [127]:
arr3200

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 3200 - FA 17 - WARREN.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 3200 - FA 17B - WARREN.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2019-2020\\Fall 2019/OMIS3200-92357-FA19-WARREN.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2019-2020\\Fall 2019/OMIS3200-92358-FA19-WARREN.pdf']

In [128]:
arr2355 = checkcrs('2355', 'OMIS')

In [129]:
arr2355

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS 2355-2017Fall.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 OMIS2355-2017Fall.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2018-19\\Fall 2018/Syllabus OMIS 2355 -2.pdf']

**Doesn't capture scanned pdfs**

In [130]:
arr2392 = checkcrs('2392', 'OMIS')

In [131]:
arr2392

[]

**Doesn't capture scanned pdfs**

In [132]:
arr2627 = checkcrs ('2627', 'MSIS')

In [133]:
arr2627

[]

In [107]:
#dfcourse['check'] = dfcourse['Catalog'].apply(checkcrs)

**Prereq ('3000') mentioned on the first page of the syllabus hence, syllabus for 3250 also got captured**

In [141]:
arr3000 = checkcrs ('3000', 'OMIS')

In [142]:
arr3000

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 Syllabus OMIS3000.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2018-19\\Fall 2018/1Course Syllabus OMIS3000.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2018-19\\Fall 2018/2018 Fall OMIS3250 Syllabus 2018.08.30.pdf']

**Some of the syllabi have MSIS2641 without a space between department and course number. Both types are captured**

In [143]:
arr2641 = checkcrs ('2641', 'MSIS')

In [144]:
arr2641

['C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2641_Fall2017_Syllabus-Seidler (1).pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Fall 2017/Fall 2017 MSIS2641_Fall2017_Syllabus-Seidler.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Spring 2018/Spring 2018 MSIS2641_Spring_2018_Syllabus-Seidler.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2017-18\\Winter 2018/Winter 2018 MSIS2641_Winter2018_Syllabus-Seidler.pdf',
 'C:\\Users\\Nachiket\\Desktop\\Capstone\\Syllabi\\2019-2020\\Fall 2019/MSIS 2641 Seidler.pdf']

In [148]:
dfcourse['path'] = dfcourse.apply(checkcrs,Catalog,Subject)

NameError: name 'Catalog' is not defined

**Overall Observations**
1. I am not sure if it can capture scanned pdfs. For the courses tested till now, they are not captured
2. Considered only 1st and 2nd page, but if prereqs are there on either of the pages they are also captured
3. Somehow not working properly on OMIS 3000. Need help figuring out why 

**Extract information from the word files**

**Approach1** - Cancelled

In [156]:
#specific to extracting information from word documents
import os
import zipfile
#other tools useful in extracting the information from our document
import re
#to pretty print our xml:
import xml.dom.minidom

In [157]:
#os.listdir('.')

In [167]:
document = zipfile.ZipFile('C:\\Users\\Nachiket\\Desktop\\Capstone\\syllabi\\2017-18\\Fall 2017\\Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx')
#document will be the filetype zipfile.ZipFile


In [168]:
document

<zipfile.ZipFile filename='C:\\Users\\Nachiket\\Desktop\\Capstone\\syllabi\\2017-18\\Fall 2017\\Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx' mode='r'>

In [169]:
document.namelist()

['[Content_Types].xml',
 '_rels/.rels',
 'word/_rels/document.xml.rels',
 'word/document.xml',
 'word/endnotes.xml',
 'word/footnotes.xml',
 'word/footer1.xml',
 'word/footer2.xml',
 'word/theme/theme1.xml',
 'word/settings.xml',
 'word/styles.xml',
 'word/numbering.xml',
 'docProps/app.xml',
 'customXml/_rels/item1.xml.rels',
 'customXml/itemProps1.xml',
 'customXml/item1.xml',
 'word/fontTable.xml',
 'word/webSettings.xml',
 'word/stylesWithEffects.xml',
 'docProps/core.xml']

In [171]:
#name = 'word/people.xml'
#we can see who the document author is: Stephen C. Phillips
uglyXml = xml.dom.minidom.parseString(document.read('word/document.xml')).toprettyxml(indent='  ')

text_re = re.compile('>\n\s+([^<>\s].*?)\n\s+</', re.DOTALL)    
prettyXml = text_re.sub('>\g<1></', uglyXml)

print(prettyXml)

<?xml version="1.0" ?>
<w:document mc:Ignorable="w14 wp14" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordp

In [172]:
#name = 'word/fontTable.xml'
#This looks like the fonts used in the document style
uglyXml = xml.dom.minidom.parseString(document.read('word/fontTable.xml')).toprettyxml(indent='  ')

text_re = re.compile('>\n\s+([^<>\s].*?)\n\s+</', re.DOTALL)    
prettyXml = text_re.sub('>\g<1></', uglyXml)

print(prettyXml)

<?xml version="1.0" ?>
<w:fonts mc:Ignorable="w14" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml">
  <w:font w:name="Symbol">
    <w:panose1 w:val="05050102010706020507"/>
    <w:charset w:val="02"/>
    <w:family w:val="roman"/>
    <w:pitch w:val="variable"/>
    <w:sig w:csb0="80000000" w:csb1="00000000" w:usb0="00000000" w:usb1="10000000" w:usb2="00000000" w:usb3="00000000"/>
  </w:font>
  <w:font w:name="Times New Roman">
    <w:panose1 w:val="02020603050405020304"/>
    <w:charset w:val="00"/>
    <w:family w:val="roman"/>
    <w:pitch w:val="variable"/>
    <w:sig w:csb0="000001FF" w:csb1="00000000" w:usb0="E0002AFF" w:usb1="C0007841" w:usb2="00000009" w:usb3="00000000"/>
  </w:font>
  <w:font w:name="Courier New">
    <w:panose1 w:val="02070309

In [173]:
xml_content = document.read('word/document.xml')
document.close()
xml_str = str(xml_content)

In [187]:
xml_str

'b\'<?xml version="1.0" encoding="UTF-8" standalone="yes"?>\\r\\n<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microso

In [193]:
val = "MSIS2604-01"
flag = False
lst = xml_str.split()
if val in lst:
    flag = True
    

In [195]:
lst

["b'<?xml",
 'version="1.0"',
 'encoding="UTF-8"',
 'standalone="yes"?>\\r\\n<w:document',
 'xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas"',
 'xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"',
 'xmlns:o="urn:schemas-microsoft-com:office:office"',
 'xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"',
 'xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math"',
 'xmlns:v="urn:schemas-microsoft-com:vml"',
 'xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing"',
 'xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing"',
 'xmlns:w10="urn:schemas-microsoft-com:office:word"',
 'xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"',
 'xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml"',
 'xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup"',
 'xmlns:wpi="http://schemas.microsoft.com/offic

In [194]:
flag

False

In [185]:
link_list = re.findall('MSIS2604-01',xml_str)[1:]
#ink_list = [x[:-1] for x in link_list]

In [186]:
link_list

[]

**Approach2** - Cancelled

In [198]:
>>> import zipfile
>>> z = zipfile.ZipFile('C:\\Users\\Nachiket\\Desktop\\Capstone\\syllabi\\2017-18\\Fall 2017\\Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx')
>>> "MSIS2604" in z.read("word/document.xml")

>>> z.close()

TypeError: a bytes-like object is required, not 'str'

**Approach3** - Cancelled

In [203]:
from docx import *

In [204]:


# Open the .docx file
document = opendocx('C:\\Users\\Nachiket\\Desktop\\Capstone\\syllabi\\2017-18\\Fall 2017\\Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx')

# Search returns true if found    
search(document,'MSIS2604-01')

NameError: name 'opendocx' is not defined

**Approach4** - Working

In [205]:
import docx2txt
 
# read in word file
result = docx2txt.process("C:\\Users\\Nachiket\\Desktop\\Capstone\\syllabi\\2017-18\\Fall 2017\\Fall 2017 MSIS2604 Syllabus -- Fall 2017.docx")


In [206]:
result

"MSIS2604-01\n\nInformation Systems Policy & Strategy\n\nFall Quarter, 2017\n\n\n\nInstructor:\tDr. Darrel A. (Del) Mank\t\t\tdmank@scu.edu\t\t\t\t\t\t\t\t\n\n\t\tCell Phone: 408-605-3983\t\t\tOffice Hours:\tBy appointment\n\n\t\tOffice:\tRoom 321W Lucas Hall \t\tClass Days:\tTTh\t\t\n\nClass Period:\t7:35pm—8:50pm  \t\tClass Room:\tLucas Hall 306\n\n\t\n\nText:  Schilling, Melissa A.; STRATEGIC MANAGEMENT of TECHNOLOGICAL INNOVATION, McGraw-Hill/Irwin,\n\n           5th Edition, 2017           ISBN 978-1-259-53906-0\n\n\n\nCases*\t\tHewlett-Packard Merced Division [HBR]\n\n\t\tSAP AG in 2006 [Stanford]\n\n\t\tVMware Inc., 2008 [HBR]\n\n\t\tSkype vs AT&T [University of Alabama]\n\n\t\tOracle vs. salesforce.com [HBR]\n\n\t\tNetflix in 2012 [University of Alabama]\n\n\t\tGoogle Strategy in 2012 [University of South Alabama]\n\n\n\n\t\t*All Cases are available at the SCU Bookstore\n\n\n\nCourse Objectives: \n\nIdentify the range, scope, and complexity of the issues and problems related to

In [222]:
val = "MSIS2604"
flag = False
lst = xml_str.split()
if val in result:
    flag = True

In [223]:
flag

True

In [220]:
def checkcrsdocx(param,dep):
    arr = []
    val = str(param)
    dep = dep
    comb = dep+param
    for root in rootdocx:
        result = docx2txt.process(root)
        if val in result:
            if dep in result:
                arr.append(root)
        elif comb in result:
            arr.append(root)
    return arr

In [221]:
arr2604 = checkcrsdocx('2604','MSIS')

BadZipFile: File is not a zip file